File: python/agento11y/experiments/evaluators.py
"""Locally configured evaluator helpers for experiment trials."""

from __future__ import annotations

import json
import re
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any, Protocol

from ..models import TokenUsage
from .types import Evaluator, EvaluatorKind

JudgeInvoke = Callable[[str], Any]
JudgeParser = Callable[[str], tuple[float, bool, str]]
JudgeUsageExtractor = Callable[[Any], TokenUsage | None]

DEFAULT_LLM_JUDGE_PROMPT = """Grade the candidate output against the input and expected result.

Input:
{input}

Expected:
{expected}

Candidate output:
{output}

Return only JSON with this shape:
{"score": <number from 0 to 1>, "passed": <boolean>, "explanation": "<brief reason>"}
"""


@dataclass(frozen=True, slots=True)
class GraderGeneration:
    """A grader request/response that can be published as a generation."""

    input: str
    output: str
    model_provider: str
    model_name: str
    agent_name: str = "agento11y-llm-judge"
    agent_version: str = ""
    operation_name: str = "evaluate"
    usage: TokenUsage | None = None


@dataclass(frozen=True, slots=True)
class EvaluationResult:
    """A local evaluator result ready to attach to a trial."""

    evaluator: Evaluator
    value: float | bool | str
    passed: bool
    explanation: str = ""
    score_key: str = "final"
    metadata: dict[str, Any] = field(default_factory=dict)
    grader: GraderGeneration | None = None


class OutputEvaluator(Protocol):
    """Evaluator that scores caller-supplied output without a platform definition."""

    evaluator: Evaluator

    def evaluate_output(self, *, input: Any, output: Any, expected: Any = None) -> EvaluationResult: ...


@dataclass(slots=True)
class LLMJudge:
    """Provider-neutral LLM judge with automatic structured-result parsing.

    ``invoke`` receives the rendered prompt and may return a string or an object
    with a ``content`` attribute, which covers common provider and framework
    clients. No platform evaluator or control-plane credential is required.
    """

    evaluator_id: str
    invoke: JudgeInvoke
    model_name: str
    prompt_template: str = DEFAULT_LLM_JUDGE_PROMPT
    model_provider: str = ""
    version: str = "1"
    score_key: str = "final"
    pass_threshold: float = 0.5
    parser: JudgeParser | None = None
    usage_extractor: JudgeUsageExtractor | None = None
    agent_name: str = "agento11y-llm-judge"
    agent_version: str = ""
    operation_name: str = "llm-judge"
    evaluator: Evaluator = field(init=False)

    def __post_init__(self) -> None:
        if not self.evaluator_id.strip():
            raise ValueError("evaluator_id is required")
        if not callable(self.invoke):
            raise ValueError("invoke must be callable")
        if not self.model_name.strip():
            raise ValueError("model_name is required")
        if not self.prompt_template.strip():
            raise ValueError("prompt_template is required")
        if not 0.0 <= self.pass_threshold <= 1.0:
            raise ValueError("pass_threshold must be between 0 and 1")
        if self.usage_extractor is not None and not callable(self.usage_extractor):
            raise ValueError("usage_extractor must be callable")
        self.evaluator = Evaluator(
            evaluator_id=self.evaluator_id,
            version=self.version,
            kind=EvaluatorKind.LLM_JUDGE.value,
        )

    def evaluate_output(self, *, input: Any, output: Any, expected: Any = None) -> EvaluationResult:
        """Grades explicit input/output values; it does not fetch a conversation."""

        prompt = _render_prompt(self.prompt_template, input=input, output=output, expected=expected)
        response = self.invoke(prompt)
        raw = _response_text(response)
        usage = self.usage_extractor(response) if self.usage_extractor is not None else _response_usage(response)
        score, passed, explanation = self.parser(raw) if self.parser is not None else self._parse_default(raw)
        metadata = {
            "judge_model": self.model_name,
            "judge_provider": self.model_provider,
        }
        return EvaluationResult(
            evaluator=self.evaluator,
            value=score,
            passed=passed,
            explanation=explanation,
            score_key=self.score_key,
            metadata={key: value for key, value in metadata.items() if value},
            grader=GraderGeneration(
                input=prompt,
                output=raw,
                model_provider=self.model_provider,
                model_name=self.model_name,
                agent_name=self.agent_name,
                agent_version=self.agent_version or self.version,
                operation_name=self.operation_name,
                usage=usage,
            ),
        )

    def _parse_default(self, raw: str) -> tuple[float, bool, str]:
        objects = _json_objects(raw)
        if not objects:
            raise ValueError("LLM judge response did not contain a JSON object")
        for payload in reversed(objects):
            try:
                score = max(0.0, min(1.0, float(payload["score"])))
            except (KeyError, TypeError, ValueError):
                continue
            passed = _parse_passed(payload.get("passed", payload.get("pass")), score, self.pass_threshold)
            explanation = str(payload.get("explanation", payload.get("reason", ""))).strip()
            return score, passed, explanation
        raise ValueError("LLM judge response requires a numeric 'score'")


@dataclass(slots=True)
class RegexJudge:
    """Deterministic regular-expression evaluator for candidate output."""

    evaluator_id: str
    pattern: str
    version: str = "1"
    score_key: str = "regex_match"
    flags: int = 0
    full_match: bool = False
    negate: bool = False
    explanation: str = ""
    evaluator: Evaluator = field(init=False)
    _compiled: re.Pattern[str] = field(init=False, repr=False)

    def __post_init__(self) -> None:
        if not self.evaluator_id.strip():
            raise ValueError("evaluator_id is required")
        if not self.pattern:
            raise ValueError("pattern is required")
        self._compiled = re.compile(self.pattern, self.flags)
        self.evaluator = Evaluator(
            evaluator_id=self.evaluator_id,
            version=self.version,
            kind=EvaluatorKind.DETERMINISTIC.value,
        )

    def evaluate_output(self, *, input: Any, output: Any, expected: Any = None) -> EvaluationResult:
        del input, expected
        text = str(output)
        matched = bool(self._compiled.fullmatch(text) if self.full_match else self._compiled.search(text))
        passed = not matched if self.negate else matched
        explanation = self.explanation or (
            f"output {'did not match' if self.negate else 'matched'} /{self.pattern}/"
            if passed
            else f"output {'matched excluded' if self.negate else 'did not match'} /{self.pattern}/"
        )
        return EvaluationResult(
            evaluator=self.evaluator,
            value=passed,
            passed=passed,
            explanation=explanation,
            score_key=self.score_key,
            metadata={"pattern": self.pattern},
        )


def _response_text(response: Any) -> str:
    content = getattr(response, "content", response)
    if isinstance(content, str):
        return content
    if isinstance(content, dict):
        return json.dumps(content)
    if isinstance(content, list):
        parts: list[str] = []
        for item in content:
            if isinstance(item, str):
                parts.append(item)
            elif isinstance(item, dict) and isinstance(item.get("text"), str):
                parts.append(item["text"])
            elif isinstance(getattr(item, "text", None), str):
                parts.append(item.text)
        if parts:
            return "".join(parts)
    return str(content)


def _response_usage(response: Any) -> TokenUsage | None:
    """Extracts common SDK and framework token-usage shapes without owning them."""

    response_metadata = _field(response, "response_metadata")
    candidates = [
        _field(response, "usage_metadata"),
        _field(response, "usage"),
        _field(response_metadata, "usage"),
        _field(response_metadata, "token_usage"),
    ]
    for candidate in candidates:
        if candidate is None:
            continue
        input_details = _field(candidate, "input_token_details")
        output_details = _field(candidate, "output_token_details")
        values = {
            "input_tokens": _first_int(candidate, "input_tokens", "prompt_tokens"),
            "output_tokens": _first_int(candidate, "output_tokens", "completion_tokens"),
            "total_tokens": _first_int(candidate, "total_tokens"),
            "cache_read_input_tokens": _first_int(candidate, "cache_read_input_tokens", "cache_read_tokens")
            or _first_int(input_details, "cache_read", "cache_read_tokens"),
            "cache_write_input_tokens": _first_int(
                candidate,
                "cache_write_input_tokens",
                "cache_creation_input_tokens",
                "cache_creation_tokens",
            )
            or _first_int(input_details, "cache_write", "cache_creation", "cache_creation_tokens"),
            "reasoning_tokens": _first_int(candidate, "reasoning_tokens")
            or _first_int(output_details, "reasoning", "reasoning_tokens"),
        }
        if any(value is not None for value in values.values()):
            return TokenUsage(**{key: value or 0 for key, value in values.items()}).normalize()
    return None


def _json_objects(raw: str) -> list[dict[str, Any]]:
    """Extracts complete JSON objects without merging unrelated brace ranges."""

    decoder = json.JSONDecoder()
    objects: list[dict[str, Any]] = []
    cursor = 0
    while True:
        start = raw.find("{", cursor)
        if start < 0:
            break
        try:
            value, end = decoder.raw_decode(raw, start)
        except json.JSONDecodeError:
            cursor = start + 1
            continue
        if isinstance(value, dict):
            objects.append(value)
        cursor = max(end, start + 1)
    return objects


def _field(value: Any, name: str) -> Any:
    if isinstance(value, dict):
        return value.get(name)
    return getattr(value, name, None)


def _first_int(value: Any, *names: str) -> int | None:
    for name in names:
        raw = _field(value, name)
        if raw is None or isinstance(raw, bool):
            continue
        try:
            parsed = int(raw)
        except (TypeError, ValueError):
            continue
        if parsed >= 0:
            return parsed
    return None


def _render_prompt(template: str, *, input: Any, output: Any, expected: Any) -> str:
    """Replaces judge placeholders without treating JSON braces as formatting."""

    values = {"input": str(input), "output": str(output), "expected": str(expected)}
    return re.sub(r"\{(input|output|expected)\}", lambda match: values[match.group(1)], template)


def _parse_passed(value: Any, score: float, threshold: float) -> bool:
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        normalized = value.strip().lower()
        if normalized in {"true", "yes", "1", "pass", "passed"}:
            return True
        if normalized in {"false", "no", "0", "fail", "failed"}:
            return False
    return score >= threshold

File: js/src/experiments/evaluators.ts
import type { TokenUsage } from '../types.js';
import type { Evaluator } from './models.js';

export const DEFAULT_LLM_JUDGE_PROMPT = `Grade the candidate output against the input and expected result.

Input:
{input}

Expected:
{expected}

Candidate output:
{output}

Return only JSON with this shape:
{"score": <number from 0 to 1>, "passed": <boolean>, "explanation": "<brief reason>"}
`;

/** A grader request and response that can be published as a generation. */
export interface GraderGeneration {
  input: string;
  output: string;
  modelProvider: string;
  modelName: string;
  agentName?: string;
  agentVersion?: string;
  operationName?: string;
  usage?: TokenUsage;
}

/** A local evaluator result ready to attach to a trial. */
export interface EvaluationResult {
  evaluator: Evaluator;
  value: number | boolean | string;
  passed: boolean;
  explanation?: string;
  scoreKey?: string;
  metadata?: Record<string, unknown>;
  grader?: GraderGeneration;
}

export interface EvaluateOutputInput {
  input: unknown;
  output: unknown;
  expected?: unknown;
}

/** Evaluator that scores caller-supplied output without a platform definition. */
export interface OutputEvaluator {
  readonly evaluator: Evaluator;
  evaluateOutput(input: EvaluateOutputInput): Promise<EvaluationResult> | EvaluationResult;
}

/** The parsed fields a judge response has to yield. */
export interface ParsedJudgeResult {
  score: number;
  passed: boolean;
  explanation: string;
}

export interface LLMJudgeOptions {
  evaluatorId: string;
  /** Receives the rendered prompt. May return a string or a provider response object. */
  invoke: (prompt: string) => Promise<unknown> | unknown;
  modelName: string;
  promptTemplate?: string;
  modelProvider?: string;
  version?: string;
  scoreKey?: string;
  passThreshold?: number;
  /** Replaces the default lenient JSON parsing. */
  parser?: (raw: string) => ParsedJudgeResult;
  /** Replaces the default token-usage extraction. */
  usageExtractor?: (response: unknown) => TokenUsage | undefined;
  agentName?: string;
  agentVersion?: string;
  operationName?: string;
}

/**
 * Provider-neutral LLM judge with structured-result parsing.
 *
 * `invoke` receives the rendered prompt and may return a string or an object with
 * a `content` field, which covers common provider and framework clients. No
 * platform evaluator or control-plane credential is required.
 */
export class LLMJudge implements OutputEvaluator {
  readonly evaluator: Evaluator;
  readonly scoreKey: string;
  readonly passThreshold: number;

  private readonly invoke: (prompt: string) => Promise<unknown> | unknown;
  private readonly modelName: string;
  private readonly modelProvider: string;
  private readonly promptTemplate: string;
  private readonly parser: ((raw: string) => ParsedJudgeResult) | undefined;
  private readonly usageExtractor: ((response: unknown) => TokenUsage | undefined) | undefined;
  private readonly agentName: string;
  private readonly agentVersion: string;
  private readonly operationName: string;

  constructor(options: LLMJudgeOptions) {
    if (options.evaluatorId.trim().length === 0) {
      throw new Error('evaluatorId is required');
    }
    if (typeof options.invoke !== 'function') {
      throw new Error('invoke must be callable');
    }
    if (options.modelName.trim().length === 0) {
      throw new Error('modelName is required');
    }
    const promptTemplate = options.promptTemplate ?? DEFAULT_LLM_JUDGE_PROMPT;
    if (promptTemplate.trim().length === 0) {
      throw new Error('promptTemplate is required');
    }
    const passThreshold = options.passThreshold ?? 0.5;
    if (!(passThreshold >= 0 && passThreshold <= 1)) {
      throw new Error('passThreshold must be between 0 and 1');
    }
    const version = options.version ?? '1';
    this.evaluator = { evaluatorId: options.evaluatorId, version, kind: 'llm_judge' };
    this.invoke = options.invoke;
    this.modelName = options.modelName;
    this.modelProvider = options.modelProvider ?? '';
    this.promptTemplate = promptTemplate;
    this.scoreKey = options.scoreKey ?? 'final';
    this.passThreshold = passThreshold;
    this.parser = options.parser;
    this.usageExtractor = options.usageExtractor;
    this.agentName = options.agentName ?? 'agento11y-llm-judge';
    this.agentVersion = options.agentVersion ?? version;
    this.operationName = options.operationName ?? 'llm-judge';
  }

  /** Grades explicit input and output values; it does not fetch a conversation. */
  async evaluateOutput(input: EvaluateOutputInput): Promise<EvaluationResult> {
    const prompt = renderPrompt(this.promptTemplate, input);
    const response = await this.invoke(prompt);
    const raw = responseText(response);
    const usage = this.usageExtractor !== undefined ? this.usageExtractor(response) : responseUsage(response);
    const parsed = this.parser !== undefined ? this.parser(raw) : this.parseDefault(raw);
    const metadata: Record<string, unknown> = { judge_model: this.modelName };
    if (this.modelProvider.length > 0) {
      metadata.judge_provider = this.modelProvider;
    }
    return {
      evaluator: this.evaluator,
      value: parsed.score,
      passed: parsed.passed,
      explanation: parsed.explanation,
      scoreKey: this.scoreKey,
      metadata,
      grader: {
        input: prompt,
        output: raw,
        modelProvider: this.modelProvider,
        modelName: this.modelName,
        agentName: this.agentName,
        agentVersion: this.agentVersion,
        operationName: this.operationName,
        ...(usage !== undefined ? { usage } : {}),
      },
    };
  }

  /**
   * Reads the last JSON object in the response that carries a numeric `score`.
   *
   * Judges tend to narrate before the verdict, so unrelated leading JSON is
   * skipped and the last scored object wins. A nested rubric score never replaces
   * a top-level one, because the top-level object is the outermost match.
   */
  private parseDefault(raw: string): ParsedJudgeResult {
    const objects = jsonObjects(raw);
    if (objects.length === 0) {
      throw new Error('LLM judge response did not contain a JSON object');
    }
    for (let index = objects.length - 1; index >= 0; index--) {
      const payload = objects[index];
      if (payload === undefined) {
        continue;
      }
      const rawScore = payload.score;
      const numeric = typeof rawScore === 'number' ? rawScore : Number(rawScore);
      if (rawScore === undefined || rawScore === null || typeof rawScore === 'boolean' || !Number.isFinite(numeric)) {
        continue;
      }
      const score = Math.max(0, Math.min(1, numeric));
      const passed = parsePassed(payload.passed ?? payload.pass, score, this.passThreshold);
      const explanation = String(payload.explanation ?? payload.reason ?? '').trim();
      return { score, passed, explanation };
    }
    throw new Error("LLM judge response requires a numeric 'score'");
  }
}

export interface RegexJudgeOptions {
  evaluatorId: string;
  pattern: string;
  version?: string;
  scoreKey?: string;
  flags?: string;
  fullMatch?: boolean;
  negate?: boolean;
  explanation?: string;
}

/** Deterministic regular-expression evaluator for candidate output. */
export class RegexJudge implements OutputEvaluator {
  readonly evaluator: Evaluator;
  readonly scoreKey: string;

  private readonly pattern: string;
  private readonly compiled: RegExp;
  private readonly fullMatch: boolean;
  private readonly negate: boolean;
  private readonly explanationOverride: string;

  constructor(options: RegexJudgeOptions) {
    if (options.evaluatorId.trim().length === 0) {
      throw new Error('evaluatorId is required');
    }
    if (options.pattern.length === 0) {
      throw new Error('pattern is required');
    }
    this.pattern = options.pattern;
    this.fullMatch = options.fullMatch ?? false;
    // Anchoring and the flag set are both fixed here, so the regex is compiled once
    // and reused for every evaluation, as Python's copy does. `g` is dropped
    // because `test` on a global regex advances `lastIndex` between calls.
    const source = this.fullMatch ? `^(?:${options.pattern})$` : options.pattern;
    this.compiled = new RegExp(source, (options.flags ?? '').replace(/g/g, ''));
    this.negate = options.negate ?? false;
    this.explanationOverride = options.explanation ?? '';
    this.scoreKey = options.scoreKey ?? 'regex_match';
    this.evaluator = {
      evaluatorId: options.evaluatorId,
      version: options.version ?? '1',
      kind: 'deterministic',
    };
  }

  evaluateOutput(input: EvaluateOutputInput): EvaluationResult {
    const text = String(input.output);
    const matched = this.compiled.test(text);
    const passed = this.negate ? !matched : matched;
    const explanation =
      this.explanationOverride.length > 0
        ? this.explanationOverride
        : passed
          ? `output ${this.negate ? 'did not match' : 'matched'} /${this.pattern}/`
          : `output ${this.negate ? 'matched excluded' : 'did not match'} /${this.pattern}/`;
    return {
      evaluator: this.evaluator,
      value: passed,
      passed,
      explanation,
      scoreKey: this.scoreKey,
      metadata: { pattern: this.pattern },
    };
  }
}

/** Replaces judge placeholders without treating JSON braces as formatting. */
function renderPrompt(template: string, values: EvaluateOutputInput): string {
  const replacements: Record<string, string> = {
    input: String(values.input),
    output: String(values.output),
    expected: String(values.expected),
  };
  return template.replace(/\{(input|output|expected)\}/g, (_match, key: string) => replacements[key] ?? '');
}

function responseText(response: unknown): string {
  const nested = field(response, 'content');
  const content = nested !== undefined ? nested : response;
  if (typeof content === 'string') {
    return content;
  }
  if (Array.isArray(content)) {
    const parts: string[] = [];
    for (const item of content) {
      if (typeof item === 'string') {
        parts.push(item);
      } else if (isRecordLike(item) && typeof (item as { text?: unknown }).text === 'string') {
        parts.push((item as { text: string }).text);
      }
    }
    if (parts.length > 0) {
      return parts.join('');
    }
  }
  if (isRecordLike(content)) {
    return JSON.stringify(content);
  }
  return String(content);
}

/** Extracts common SDK and framework token-usage shapes without owning them. */
function responseUsage(response: unknown): TokenUsage | undefined {
  const responseMetadata = field(response, 'response_metadata') ?? field(response, 'responseMetadata');
  const candidates = [
    field(response, 'usage_metadata') ?? field(response, 'usageMetadata'),
    field(response, 'usage'),
    field(responseMetadata, 'usage'),
    field(responseMetadata, 'token_usage') ?? field(responseMetadata, 'tokenUsage'),
  ];
  for (const candidate of candidates) {
    if (candidate === undefined || candidate === null) {
      continue;
    }
    const inputDetails = field(candidate, 'input_token_details') ?? field(candidate, 'inputTokenDetails');
    const outputDetails = field(candidate, 'output_token_details') ?? field(candidate, 'outputTokenDetails');
    const usage: TokenUsage = {
      inputTokens: firstInt(candidate, 'input_tokens', 'inputTokens', 'prompt_tokens', 'promptTokens'),
      outputTokens: firstInt(candidate, 'output_tokens', 'outputTokens', 'completion_tokens', 'completionTokens'),
      totalTokens: firstInt(candidate, 'total_tokens', 'totalTokens'),
      cacheReadInputTokens:
        firstInt(candidate, 'cache_read_input_tokens', 'cacheReadInputTokens', 'cache_read_tokens') ??
        firstInt(inputDetails, 'cache_read', 'cache_read_tokens'),
      // The field is cache_write_input_tokens, never cache_creation_input_tokens;
      // the provider spellings below are only read, not emitted.
      cacheWriteInputTokens:
        firstInt(
          candidate,
          'cache_write_input_tokens',
          'cacheWriteInputTokens',
          'cache_creation_input_tokens',
          'cache_creation_tokens',
        ) ?? firstInt(inputDetails, 'cache_write', 'cache_creation', 'cache_creation_tokens'),
      reasoningTokens:
        firstInt(candidate, 'reasoning_tokens', 'reasoningTokens') ??
        firstInt(outputDetails, 'reasoning', 'reasoning_tokens'),
    };
    if (Object.values(usage).some((value) => value !== undefined)) {
      return normalizeUsage(usage);
    }
  }
  return undefined;
}

function normalizeUsage(usage: TokenUsage): TokenUsage {
  const inputTokens = usage.inputTokens ?? 0;
  const outputTokens = usage.outputTokens ?? 0;
  return {
    inputTokens,
    outputTokens,
    totalTokens:
      usage.totalTokens !== undefined && usage.totalTokens > 0 ? usage.totalTokens : inputTokens + outputTokens,
    cacheReadInputTokens: usage.cacheReadInputTokens ?? 0,
    cacheWriteInputTokens: usage.cacheWriteInputTokens ?? 0,
    reasoningTokens: usage.reasoningTokens ?? 0,
  };
}

/**
 * Extracts complete JSON objects without merging unrelated brace ranges.
 *
 * A judge that emits two objects, or prose containing braces, would otherwise
 * produce one unparseable slice.
 */
export function jsonObjects(raw: string): Record<string, unknown>[] {
  const objects: Record<string, unknown>[] = [];
  let cursor = 0;
  while (cursor < raw.length) {
    const start = raw.indexOf('{', cursor);
    if (start < 0) {
      break;
    }
    const end = matchingBrace(raw, start);
    if (end < 0) {
      cursor = start + 1;
      continue;
    }
    try {
      const value = JSON.parse(raw.slice(start, end + 1));
      if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
        objects.push(value as Record<string, unknown>);
      }
      cursor = end + 1;
    } catch {
      cursor = start + 1;
    }
  }
  return objects;
}

/** Finds the brace closing the object at `start`, skipping braces inside strings. */
function matchingBrace(raw: string, start: number): number {
  let depth = 0;
  let inString = false;
  let escaped = false;
  for (let index = start; index < raw.length; index++) {
    const char = raw[index];
    if (inString) {
      if (escaped) {
        escaped = false;
      } else if (char === '\\') {
        escaped = true;
      } else if (char === '"') {
        inString = false;
      }
      continue;
    }
    if (char === '"') {
      inString = true;
    } else if (char === '{') {
      depth++;
    } else if (char === '}') {
      depth--;
      if (depth === 0) {
        return index;
      }
    }
  }
  return -1;
}

function parsePassed(value: unknown, score: number, threshold: number): boolean {
  if (typeof value === 'boolean') {
    return value;
  }
  if (typeof value === 'string') {
    const normalized = value.trim().toLowerCase();
    if (['true', 'yes', '1', 'pass', 'passed'].includes(normalized)) {
      return true;
    }
    if (['false', 'no', '0', 'fail', 'failed'].includes(normalized)) {
      return false;
    }
  }
  return score >= threshold;
}

function field(value: unknown, name: string): unknown {
  if (value === null || value === undefined) {
    return undefined;
  }
  if (typeof value === 'object') {
    return (value as Record<string, unknown>)[name];
  }
  return undefined;
}

function firstInt(value: unknown, ...names: string[]): number | undefined {
  for (const name of names) {
    const raw = field(value, name);
    if (raw === undefined || raw === null || typeof raw === 'boolean') {
      continue;
    }
    const parsed = typeof raw === 'number' ? Math.trunc(raw) : Number.parseInt(String(raw), 10);
    if (Number.isFinite(parsed) && parsed >= 0) {
      return parsed;
    }
  }
  return undefined;
}

function isRecordLike(value: unknown): boolean {
  return typeof value === 'object' && value !== null;
}
